// @icedesign/layout/lib/Aside.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var IceLayoutAside =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(IceLayoutAside, _Component);

  function IceLayoutAside() {
    (0, _classCallCheck2.default)(this, IceLayoutAside);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(IceLayoutAside).apply(this, arguments));
  }

  (0, _createClass2.default)(IceLayoutAside, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          type = _this$props.type,
          children = _this$props.children,
          className = _this$props.className,
          scrollable = _this$props.scrollable,
          style = _this$props.style,
          others = (0, _objectWithoutProperties2.default)(_this$props, ["type", "children", "className", "scrollable", "style"]);
      var classes = (0, _classnames.default)('ice-layout-aside', "ice-layout-aside-".concat(type), {
        'ice-layout-scrollable': scrollable
      }, className);
      return _react.default.createElement("div", (0, _extends2.default)({}, others, {
        className: classes,
        style: style
      }), children);
    }
  }]);
  return IceLayoutAside;
}(_react.Component);

exports.default = IceLayoutAside;
IceLayoutAside.displayName = 'IceLayoutAside';
IceLayoutAside.propTypes = {
  /**
   * 可滚动
   * @description Layout 开启 fixable 后有效
   */
  scrollable: _propTypes.default.bool,

  /**
   * 类型，背景色
   */
  type: _propTypes.default.oneOf(['none', 'normal', 'primary', 'secondary', 'line'])
};
IceLayoutAside.defaultProps = {
  scrollable: false,
  type: 'normal'
};
